### PR TITLE
add parsian gateway

### PR DESCRIPTION
**Pull request by mdoulabi1**

### CODE REVIEW

File: src/Drivers/Parsian/Parsian.php [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    }
+
+    private function getRequestHeaders(): array
+    {
+        return config('gateway_parsian.request_headers');
+    }
+
+    private function getSoapOptions(): array
+    {
+        return config('gateway_mellat.soap_options', [
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:463]: gateway_mellat?

File: src/Drivers/Parsian/Parsian.php [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+        return array_key_exists($statusCode, $messages) ? $messages[$statusCode] : $statusCode;
+    }
+
+    protected function getSuccessResponseStatusCode(): int
+    {
+        return  0;
+    }
+
+    protected function getPurchaseUrl(): string
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:431]: this is not used

File: src/Drivers/Parsian/Parsian.php [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    {
+        return $this->getBaseRestServiceUrl().'generateTransactionDataToSign/';
+    }
+
+    protected function getTokenGenerationUrl(): string
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:436]: i think this should be purchase url

File: src/Drivers/Parsian/Parsian.php [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+            '-107' => 'ﻗﺎﺑﻠﻴﺖ ﺍﺭﺳﺎﻝ ﺗﺎﻳﻴﺪﻩ ﺗﺮﺍﻛﻨﺶ ﺑﺮﺍﻱ ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮ ﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-106' => 'ﻗﺎﺑﻠﻴﺖ ﺷﺎﺭﮊ ﺑﺮﺍﻱ ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮ ﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-105' => 'ﻗﺎﺑﻠﻴﺖ ﺗﺎﭖ ﺁﭖ ﺑﺮﺍﻱ ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮ ﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-104' => 'ﻗﺎﺑﻠﻴﺖ ﭘﺮﺩﺍﺧﺖ ﻗﺒﺾ ﺑﺮﺍﻱ ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮ ﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-103' => 'ﻗﺎﺑﻠﻴﺖ ﺧﺮﻳﺪ ﺑﺮﺍﻱ ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮ ﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-102' => 'ﺗﺮﺍﻛﻨﺶ ﺑﺎ ﻣﻮﻓﻘﻴﺖ ﺑﺮﮔﺸﺖ ﺩﺍﺩﻩ ﺷﺪ 	',
+            '-101' => 'ﭘﺬﻳﺮﻧﺪﻩ ﺍﻫﺮﺍﺯ ﻫﻮﻳﺖ ﻧﺸﺪ 	',
+            '-100' => 'ﭘﺬﻳﺮﻧﺪﻩ ﻏﻴﺮﻓﻌﺎﻝ ﻣﻲ ﺑﺎﺷﺪ 	',
+            '-1' => 'ﺧﻄﺎﻱ ﺳﺮﻭﺭ 	',
+            '0' => 'ﻣﻮﻓﻖ 	',
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:330]: lots of spaces in messages, trim them

File: src/Drivers/Parsian/Parsian.php [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    protected function getPurchaseData(): array
+    {
+        $phoneNumber = $this->getInvoice()->getPhoneNumber();
+
+        if (!empty($phoneNumber)) {
+            $phoneNumber = $this->checkPhoneNumberFormat($phoneNumber);
+        }
+
+        return [
+            'WSContext' => $this->getAuthData(),
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:102]: getAuthData? it's from another driver i think...

File: src/Drivers/Parsian/Parsian.php [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+            'TransType' => static::BANK_BUY_TRANSACTION_TYPE,
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:103]: Undefined TransType... please check input names with bank documentation

File: config/gateway_parsian.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    /**
+     *  driver class namespace
+     */
+    'driver' => Omalizadeh\MultiPayment\Drivers\Parsian\Parsian::class,
+
+    /**
+     *  gateway configurations
+     */
+    'main' => [
+        'merchant'  => '', // use 'zibal' for sandbox (testing) mode
```
Comment (omalizadeh) [inline config/gateway_parsian.php:14]: no merchant needed. where is pin?

File: config/gateway_parsian.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        'callback_url' => 'https://yoursite.com/path/to',
+        'description' => 'payment using zarinpal',
+    ],
+    'other' => [
+        'merchant'  => '',
+        'callback_url' => 'https://yoursite.com/path/to',
+        'description' => 'payment using zarinpal',
```
Comment (omalizadeh) [inline config/gateway_parsian.php:21]: payment using parsian

File: src/Drivers/Parsian/Parsian.php [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+    }
+
+    private function getBaseRestServiceUrl(): string
+    {
+        return 'https://pec.shaparak.ir/';
+    }
+
+    private function getRequestHeaders(): array
+    {
+        return config('gateway_parsian.request_headers');
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:458]: remove this method and headers from config file. bank gateway is using SOAP

File: src/Drivers/Parsian/Parsian.php [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+    {
+        return 'https://pec.shaparak.ir/NewIPG/?token='.$this->getInvoice()->getTransactionId();
+    }
+
+    protected function getVerificationUrl(): string
+    {
+        return $this->getBaseRestServiceUrl().'NewIPGServices/Confirm/ConfirmService.asmx?wsdl';
+    }
+
+    private function getBaseRestServiceUrl(): string
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:451]: it's not rest

**Review by omalizadeh — CHANGES_REQUESTED** [review]
Please re-check documentation @mdoulabi1

File: src/Drivers/Parsian/Parsian.php (hunk 2 above)
Comment (mdoulabi1) [reply src/Drivers/Parsian/Parsian.php:431]: this method is abstract and it's not possible to remove

File: src/Drivers/Parsian/Parsian.php (hunk 6 above)
Comment (mdoulabi1) [reply src/Drivers/Parsian/Parsian.php:103]: fix

File: config/gateway_parsian.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    'soap_options' => [
+        'encoding' => 'UTF-8',
+    ],
+
+    /**
+     *  gateway configurations
+     */
+    'behandam' => [
+        'terminal_id' => '',
```
Comment (omalizadeh) [inline config/gateway_parsian.php:21]: there is no terminal_id in parsian gateway. remove this key.

File: config/gateway_parsian.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        'pin_code' => '',
+        'callback_url' => 'https://yoursite.com/path/to',
+        'callback_method' => 'POST', // Supported values: POST, GET
```
Comment (omalizadeh) [inline config/gateway_parsian.php:24]: does this gateway support get and post method?

File: config/gateway_parsian.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        'description' => 'رژیم دکتر کرمانی',
```
Comment (omalizadeh) [inline config/gateway_parsian.php:25]: Use general strings and messages

File: config/gateway_parsian.php [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    ],
+
+    'other' => [
+        'merchant'  => '',
```
Comment (omalizadeh) [inline config/gateway_parsian.php:29]: pin, not merchant

File: config/gateway_parsian.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+     *  soap client options
+     */
+    'soap_options' => [
+        'encoding' => 'UTF-8',
+    ],
+
+    /**
+     *  gateway configurations
+     */
+    'behandam' => [
```
Comment (omalizadeh) [inline config/gateway_parsian.php:20]: main

File: src/Drivers/Parsian/Parsian.php [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+        $token=$verificationResponse->ConfirmPaymentResult->Token;
+        $cardNumber=$verificationResponse->ConfirmPaymentResult->CardNumberMasked;
+        if ($verificationResponse->ConfirmPaymentResult->Status === $this->getSuccessResponseStatusCode() && $RRN > 0) {
+            return new Receipt($this->getInvoice(), $RRN, $token, $cardNumber);
+        }
+    }
+
+    protected function getPurchaseData(): array
+    {
+        if (empty($this->settings['terminal_id'])) {
```
Comment (omalizadeh) [inline src/Drivers/Parsian/Parsian.php:54]: no terminal_id needed

**Review by omalizadeh — APPROVED** [review]

**Pull request merged**